borg/archiver.py

```diff
@@ -406,64 +406,127 @@ def item_is_hardlink_master(item):
     def do_diff(self, args):
         """Diff contents of two archives"""
         def format_bytes(count):
-            if count is None:
-                return "<deleted>"
-            return format_file_size(count)
+            if count > 0:
+                sign = '+'
+            elif count < 0:
+                sign = '-'
+            else:
+                sign = ''
+
+            return '{}{}'.format(sign, format_file_size(abs(count)))
 
         def fetch_and_compare_chunks(chunk_ids1, chunk_ids2, archive1, archive2):
             chunks1 = archive1.pipeline.fetch_many(chunk_ids1)
             chunks2 = archive2.pipeline.fetch_many(chunk_ids2)
             return self.compare_chunk_contents(chunks1, chunks2)
 
+        def sum_chunk_size(item):
+            if item.get(b'deleted'):
+                return None
+            else:
+                return sum(c[1] for c in item[b'chunks'])
+
         def get_owner(item):
             if args.numeric_owner:
                 return item[b'uid'], item[b'gid']
             else:
                 return item[b'user'], item[b'group']
 
+        def get_mode(item):
+            mode = stat.filemode(item[b'mode'])
+
+            # Special notation for hard links
+            if mode[0] == '-' and b'source' in item:
+                mode = "h" + mode[1:]
```

Comment:
hmm, not sure that is helpful.

"hardlinks" are just additional names pointing to same inode. there is no difference in principle between first name and additional names in the filesystem.

in a borg archive, there is a difference: first is with data, 2nd+ is without data, pointing backwards to first one.
but that is only to save storing the data again and also to trigger the hardlinking in the fs.

but we should not display the first name as "normal file" and the 2nd+ as "hardlink", that gives an asymetric impression of the symetric filesystem reality.